ethereum/watcher/watcher.go

```diff
@@ -0,0 +1,99 @@
+package watcher
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	eth "github.com/ethereum/go-ethereum"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+var checkForBlocksTimeout = time.Second
+
+// EventFilter filters the chain for specific events (logs).
+// When it finds a desired log, it puts it into its outbound channel.
+type EventFilter struct {
+	ctx         context.Context
+	ec          *ethclient.Client
+	topic       ethcommon.Hash
+	filterQuery eth.FilterQuery
+	logCh       chan<- ethtypes.Log
+}
+
+// NewEventFilter returns a new *EventFilter.
+func NewEventFilter(
+	ctx context.Context,
+	ec *ethclient.Client,
+	contract ethcommon.Address,
+	fromBlock *big.Int,
+	topic ethcommon.Hash,
+	logCh chan<- ethtypes.Log,
+) *EventFilter {
+	filterQuery := eth.FilterQuery{
+		FromBlock: fromBlock,
+		Addresses: []ethcommon.Address{contract},
+	}
+
+	return &EventFilter{
+		ctx:         ctx,
+		ec:          ec,
+		topic:       topic,
+		filterQuery: filterQuery,
+		logCh:       logCh,
+	}
+}
+
+// Start starts the EventFilter. It watches the chain for logs.
+func (f *EventFilter) Start() error {
+	header, err := f.ec.HeaderByNumber(f.ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-f.ctx.Done():
+				return
+			case <-time.After(checkForBlocksTimeout):
+			}
+
+			currHeader, err := f.ec.HeaderByNumber(f.ctx, nil)
+			if err != nil {
+				continue
+			}
+
+			if currHeader.Number.Cmp(header.Number) <= 0 {
+				// no new blocks, don't do anything
+				continue
+			}
+
+			// let's see if we have logs
+			logs, err := f.ec.FilterLogs(f.ctx, f.filterQuery)
+			if err != nil {
+				continue
+			}
+
+			for _, l := range logs {
+				if l.Topics[0] != f.topic {
+					continue
+				}
+
+				if l.Removed {
+					continue
+				}
+
+				f.logCh <- l
+			}
+
+			// the filter inclusive of the latest block when `ToBlock` is nil, so we add 1
+			f.filterQuery.FromBlock = big.NewInt(0).Add(currHeader.Number, big.NewInt(1))
```

Comment:
In the comment, you probably wanted to write `The filter includes the ...`.  Can we rename `header` to `prevHeader`? Don't we want to be adding +1 to the previous header and not the current header to ensure that we never skip a block? Possible scenarios I'm thinking of are:
* Pushing onto the log channel blocks for an extended period
* 2 blocks are created while we are sleeping
* A new block is created after we get the header, but before we execute the filter

Reply:
updated the comment!

I think the loop is fine, since the filter query `ToBlock` is nil so it auto-sets to the current block number, and is also inclusive on that block. so for example if it queries blocks 1-4, `currHeader.Number` is 4, then the next query will have `FromBlock = 5` up until and including whatever the new highest block is

Reply:
I understand why it works now.  I was proposing that we increment from `prevHeader` (what I'll call `header`), but your code is already doing exactly that. Your initializing `FromBlock` at the end of the loop where `curHeader`, for practical purposes, is `prevHeader`.

Was the goal of doing it this way, instead of initializing `FromHeader` before the filter operation, to have `FromHeader` be `nil` on the 1st loop iteration? Setting `FromHeader` to `nil` searches from the genesis block.  Will that search be fast enough on mainnet?

Reply:
there isn't a reason to search the logs from genesis - right now it searches from the block which the swap starts at (see swap_state.go newSwapState for both maker/taker). so FromBlock on the first iteration is the start block of the swap